### PR TITLE
filter most recent report by field values.

### PR DESCRIPTION
**Pull request by benkags**

### CODE REVIEW

File: src/web/nootils.js [hunk 1]
```diff
@@ -31,16 +31,17 @@ module.exports = function(settings) {
...
     },
-    getMostRecentReport: function(reports, form) {
+    getMostRecentReport: function(reports, form, fields=null) {
       var result = null;
       reports.forEach(function(report) {
         if (report.form === form &&
            !report.deleted &&
-           (!result || report.reported_date > result.reported_date)) {
+           (!result || report.reported_date > result.reported_date) &&
+           (!fields || report.fields && _.reduce(_.keys(fields), function(memo, e) { return memo && _.propertyOf(report.fields)(e.split('.')) === fields[e]; }, true))) {
```
Comment (kennsippell) [inline src/web/nootils.js]: what is the thinking to name this `memo`? can we avoid single letter variables like `e`?

Comment (kennsippell) [inline src/web/nootils.js]: seems this line would be best split up. Maybe `Object.keys(fields).reduce((agg, field) => agg && doesFieldMatch(field), true)`?

File: test/nootils.spec.js [hunk 2]
```diff
@@ -142,3 +142,25 @@ exports['getMostRecentReport ignores deleted reports'] = function(test) {
...
   test.done();
 };
+
+exports['getMostRecentReport returns report matching fields'] = function(test) {
+  var reports = [
+    { _id: 1, form: 'H', reported_date: 1 },
+    { _id: 2, form: 'V', reported_date: 2, fields: { dob: '2000-01-01', screening: { malaria: false } }},
+    { _id: 3, form: 'V', reported_date: 3 }
+  ];
+  var actual = nootils.getMostRecentReport(reports, 'V', { 'screening.malaria': false });
```
Comment (kennsippell) [inline test/nootils.spec.js]: would it be valuable to have a test for matching multiple fields?

File: src/web/nootils.js [hunk 3]
```diff
@@ -31,16 +31,17 @@ module.exports = function(settings) {
...
       return _.findWhere(settings.tasks.schedules, { name: name });
     },
-    getMostRecentTimestamp: function(reports, form) {
-      var report = this.getMostRecentReport(reports, form);
+    getMostRecentTimestamp: function(reports, form, fields=null) {
+      var report = this.getMostRecentReport(reports, form, fields);
       return report && report.reported_date;
     },
-    getMostRecentReport: function(reports, form) {
+    getMostRecentReport: function(reports, form, fields=null) {
```
Comment (kennsippell) [inline src/web/nootils.js]: a default value of null doesn't seem to have an impact. might as well remain undefined. is it useful?

File: src/web/nootils.js (hunk 1 above)
Comment (kennsippell) [inline src/web/nootils.js]: since this is `a || b && c` it is nice to enclose in brackets `a || (b && c)`

Comment (benkags) [reply src/web/nootils.js]: Using better names.

File: src/web/nootils.js (hunk 3 above)
Comment (benkags) [reply src/web/nootils.js]: It's not. Fixing it.

**Comment by benkags** [comment]
I've pushed a commit addressing the review commits.

File: src/web/nootils.js [hunk 4]
```diff
@@ -73,6 +73,20 @@ module.exports = function(settings) {
     isDateValid: function(date) {
       return !isNaN(date.getTime());
     },
+    getField: function(report, field) {
+      /* field is a string e.g 'dob' (equivalent to report.fields.dob) or
```
Comment (kennsippell) [inline src/web/nootils.js]: consider using the jsdoc syntax for this documentation https://devhints.io/jsdoc

File: src/web/nootils.js [hunk 5]
```diff
@@ -73,6 +73,20 @@ module.exports = function(settings) {
...
+      * 'screening.test_result' equivalent to report.fields.screening.test_result
+      */
+      return _.propertyOf(report.fields)(field.split('.'));
+    },
+    fieldEquals: function(report, field, value) {
```
Comment (kennsippell) [inline src/web/nootils.js]: is this function helpful? seems about as easy to just write `Utils.getField(report, field) === value` vs `Utils.getField(report, field, value)`

File: src/web/nootils.js [hunk 6]
```diff
@@ -31,16 +31,17 @@ module.exports = function(settings) {
     getSchedule: function(name) {
       return _.findWhere(settings.tasks.schedules, { name: name });
     },
-    getMostRecentTimestamp: function(reports, form) {
-      var report = this.getMostRecentReport(reports, form);
+    getMostRecentTimestamp: function(reports, form, fields=null) {
```
Comment (kennsippell) [inline src/web/nootils.js]: one more default value that isn't required here

File: src/web/nootils.js [hunk 7]
```diff
@@ -72,6 +73,20 @@ module.exports = function(settings) {
...
+      /* field is a string e.g 'dob' (equivalent to report.fields.dob) or
+      * 'screening.test_result' equivalent to report.fields.screening.test_result
+      */
+      return _.propertyOf(report.fields)(field.split('.'));
+    },
+    fieldEquals: function(report, field, value) {
+      return lib.getField(report, field) === value;
+    },
+    fieldsMatch: function(report, fieldValues) {
+      return Object.keys(fieldValues).reduce(function(agg, field) {
```
Comment (kennsippell) [inline src/web/nootils.js]: nit: Sorry for this late thought -- but it is also possible to write this as `Object.keys(fieldValues).every(field => lib.getField(report, field) === fieldValues(field))`. Up to you what is cleaner.

**Pull request merged**